File: Blind 75/01 - Arrays and Hashing/347-top-k-frequent-elements.py
'''
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.
'''

from collections import defaultdict
from time import time
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count_map = defaultdict(int)
        # Where i = count and max(count) <= len(nums)
        buckets = [[] for _ in range(len(nums) + 1)]
        for n in nums:
            count_map[n] += 1
        for n, c in count_map.items():
            buckets[c].append(n)

        # Start from the largest bucket
        res = []
        for i in range(len(buckets) - 1, 0, -1):
            for n in buckets[i]:
                res.append(n)
                if len(res) == k:
                    # Only add up to k
                    return res

    def reference(self, nums: List[int], k: int) -> List[int]:
        count = {}
        freq = [[] for i in range(len(nums) + 1)]

        for n in nums:
            count[n] = 1 + count.get(n, 0)
        for n, c in count.items():
            freq[c].append(n)

        res = []
        for i in range(len(freq) - 1, 0, -1):
            for n in freq[i]:
                res.append(n)
                if len(res) == k:
                    return res

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.topKFrequent(*case))
                else:
                    self.topKFrequent(*case)
        print(f'Runtime for our solution: {time() - sol_start}\n')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(*case))
                else:
                    self.reference(*case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [([1, 1, 1, 2, 2, 3], 2), ([1], 1)]
    test.quantify(test_cases)

File: Blind 75/01 - Arrays and Hashing/JavaScript/347-top-k-frequent-elements.js
const { performance } = require("perf_hooks");

class Solution {
  method(nums, k) {
    const buckets = Array.from({ length: nums.length + 1 }, () => []);
    const countMap = {};

    for (const n of nums) {
      countMap[n] = (countMap[n] || 0) + 1;
    }

    for (const [n, count] of Object.entries(countMap)) {
      buckets[count].push(n);
    }

    const res = [];
    for (let i = buckets.length - 1; i > -1; i--) {
      for (const n of buckets[i]) {
        res.push(Number(n));
        if (res.length === k) return res;
      }
    }

    return res;
  }

  reference(nums, k) {
    let map = {};
    let res = [];
    let bucket = Array.from({ length: nums.length + 1 }, () => []); // To create unique arrays

    // Storing frequency of numbers in a map
    for (const n of nums) {
      map[n] = n in map ? 1 + map[n] : 1;
    }

    // Populate the bucket with numbers in frequency
    // as the index of the bucket
    for (const c in map) {
      bucket[map[c]].push(c);
    }

    for (let i = bucket.length - 1; i >= 0; i--) {
      if (bucket[i].length > 0) {
        bucket[i].forEach((elem) => res.push(Number(elem)));
        if (k == res.length) {
          return res;
        }
      }
    }
  }

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(...input));
        else this.method(...input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(...input));
        else this.reference(...input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [[1, 1, 1, 2, 2, 3], 2],
  [[1], 1],
];
test.quantify(testCases);
